### ISSUE TITLE
Optimizers

### ISSUE THREAD

**Issue by davidmascharka** [davidmascharka/MyNN#2]
An optimizer should take an iterable of parameters at creation and perform some optimization over those parameters based on a loss that is passed to it. It can also be used to null the gradient of each one of those parameters. Something like the following:

``` python
# create a model in `model` and loss function in `loss_func`
optim = optimizers.sgd(model.parameters, learning_rate=0.01, momentum=0.99)

for batch in training_set:
    optim.null_gradients()
    outputs = model.forward(batch)
    loss = loss_func(outputs, targets)
    optim.step() # backprop into `model.parameters` and update each one
```

### The Optimizers
- [x] SGD (with momentum, maybe with [Nesterov](http://www.cs.toronto.edu/%7Ehinton/absps/momentum.pdf))
- [x] [Adam](https://arxiv.org/abs/1412.6980)
- [x] [AdaMax](https://arxiv.org/abs/1412.6980)
- [ ] [amsgrad](https://openreview.net/forum?id=ryQu7f-RZ)
- [ ] [Adadelta](https://arxiv.org/abs/1212.5701)
- [ ] [Adagrad](http://jmlr.org/papers/volume12/duchi11a/duchi11a.pdf) (PDF link)
- [ ] [(L-)BFGS](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm)
- [ ] [rmsprop](http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf) (link to Hinton lecture)
- [ ] [Adafactor](https://arxiv.org/abs/1804.04235)

**Comment by davidmascharka** [comment]
Separated these out for ease of tracking.
